diff --git a/libweston/backend-drm/kms.c b/libweston/backend-drm/kms.c
index 6f5d737d..5a5616e0 100644
--- a/libweston/backend-drm/kms.c
+++ b/libweston/backend-drm/kms.c
@@ -1122,7 +1122,7 @@ drm_pending_state_apply_atomic(struct drm_pending_state *pending_state,
 	drmModeAtomicReq *req = drmModeAtomicAlloc();
 	uint32_t flags;
 	int ret = 0;
-	drm_magic_t magic;
+//	drm_magic_t magic;
 
 	if (!req)
 		return -1;
@@ -1235,11 +1235,11 @@ drm_pending_state_apply_atomic(struct drm_pending_state *pending_state,
 		goto out;
 	}
 
-	/*drm master was set by systemd in PM test, try to set the master back.*/
-	if (!(drmGetMagic(b->drm.fd, &magic) == 0 &&
-			drmAuthMagic(b->drm.fd, magic) == 0)) {
-		drmSetMaster(b->drm.fd);
-	}
+//	/*drm master was set by systemd in PM test, try to set the master back.*/
+//	if (!(drmGetMagic(b->drm.fd, &magic) == 0 &&
+//			drmAuthMagic(b->drm.fd, magic) == 0)) {
+//		drmSetMaster(b->drm.fd);
+//	}
 	ret = drmModeAtomicCommit(b->drm.fd, req, flags, b);
 	drm_debug(b, "[atomic] drmModeAtomicCommit\n");
 
-- 
2.25.1

